File: metrics-yammer/src/main/java/ly/stealth/kafka/metrics/TopicReporter.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ly.stealth.kafka.metrics;

import com.yammer.metrics.core.*;
import com.yammer.metrics.reporting.AbstractPollingReporter;
import com.yammer.metrics.stats.Snapshot;
import kafka.javaapi.producer.Producer;
import kafka.producer.KeyedMessage;
import kafka.producer.ProducerConfig;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import static java.lang.String.format;
import static java.lang.String.valueOf;

public class TopicReporter extends AbstractPollingReporter implements MetricProcessor<Context> {
    private final Map<MetricName, Producer> producerMap = new HashMap<MetricName, Producer>();
    private final MetricPredicate predicate = MetricPredicate.ALL;
    private final Clock clock = Clock.defaultClock();
    private final ProducerConfig producerConfig;
    private final String prefix;

    private Long startTime = 0L;

    public TopicReporter(MetricsRegistry metricsRegistry, ProducerConfig producerConfig, String prefix) {
        super(metricsRegistry, "kafka-topic-reporter");
        this.producerConfig = producerConfig;
        this.prefix = prefix;
    }

    public void run() {
        final Set<Map.Entry<MetricName, Metric>> metrics=getMetricsRegistry().allMetrics().entrySet();
        try {
            for (Map.Entry<MetricName, Metric> entry : metrics) {
                final MetricName metricName = entry.getKey();
                final Metric metric = entry.getValue();
                if (predicate.matches(metricName, metric)) {
                    final Context context = new Context() {
                        public Producer getProducer() {
                            return getActualProducer(metricName);
                        }
                    };
                    metric.processWith(this, entry.getKey(), context);
                }
            }
        } catch(Exception e) {
            e.printStackTrace();
        }
    }

    public void processMeter(MetricName name, Metered meter, Context context) {
        final String header = "# time,count,1 min rate,mean rate,5 min rate,15 min rate";
        final Producer producer = context.getProducer();
        final String topic = "%s-metrics-meter".format(prefix);
        final String message = valueOf(meter.count()) + ',' + meter.oneMinuteRate() + ',' + meter.meanRate() + ','
                + meter.fiveMinuteRate() + ',' + meter.fifteenMinuteRate();
        send(producer, header, topic, message);
    }

    public void processCounter(MetricName name, Counter counter, Context context) {
        final String header = "# time,count";
        final Producer producer = context.getProducer();
        final String topic = "%s-metrics-counter".format(prefix);
        final String message = valueOf(counter.count());
        send(producer, header, topic, message);
    }

    public void processHistogram(MetricName name, Histogram histogram, Context context) {
        final String header = "# time,min,max,mean,median,stddev,95%,99%,99.9%";
        final Producer producer = context.getProducer();
        final Snapshot snapshot = histogram.getSnapshot();
        final String topic="%s-metrics-histogram".format(prefix);
        final String message = valueOf(histogram.min()) + ',' + histogram.max() + ',' + histogram.mean() + ','
                + snapshot.getMedian() + ',' + histogram.stdDev() + ',' + snapshot.get95thPercentile() + ',' + snapshot.get99thPercentile() + ','
                + snapshot.get999thPercentile();
        send(producer, header, topic, message);
    }

    public void processTimer(MetricName name, Timer timer, Context context) {
        final String header = "# time,min,max,mean,median,stddev,95%,99%,99.9%";
        final Producer producer = context.getProducer();
        final Snapshot snapshot = timer.getSnapshot();
        final String topic="%s-metrics-timer".format(prefix);
        final String  message = valueOf(timer.min()) + ',' + timer.max() + ',' + timer.mean() + ',' + snapshot.getMedian() + ','
                + timer.stdDev() + ',' + snapshot.get95thPercentile() + ',' + snapshot.get99thPercentile() + ',' + snapshot.get999thPercentile();
        send(producer, header, topic, message);
    }

    public void processGauge(MetricName name, Gauge<?> gauge, Context context){
        final String header = "# time,finalue";
        final Producer producer = context.getProducer();
        final String topic = "%s-metrics-gauge".format(prefix);
        final String message = gauge.value().toString();
        send(producer, header, topic, message);
    }

    @Override
    public void start(long period, TimeUnit unit) {
        this.startTime = clock.time();
        super.start(period, unit);
    }

    @Override
    public void shutdown() {
        try {
            super.shutdown();
        } finally {
            for (Producer producer : producerMap.values()) {
                producer.close();
            }
        }
    }

    private Producer getActualProducer(MetricName metricName) {
        Producer producer;
        synchronized(producerMap) {
            producer = producerMap.get(metricName);
            if (producer == null) {
                producer = new Producer(producerConfig);
                producerMap.put(metricName, producer);
            }
        }
        return producer;
    }

    private void send(Producer producer,String header, String topic, String message) {
        final Long time = TimeUnit.MILLISECONDS.toSeconds(clock.time() - startTime);
        try {
            producer.send(new KeyedMessage(topic, format("%s\n%d,%s", header, time, message).getBytes("UTF-8")));
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }
}

interface Context{
    public Producer getProducer();
}

File: metrics/src/main/scala/ly/stealth/kafka/core/metrics/TopicReporter.scala
package ly.stealth.kafka.core.metrics

import com.yammer.metrics.core._
import com.yammer.metrics.stats.Snapshot
import java.util.concurrent.TimeUnit
import com.yammer.metrics.reporting.AbstractPollingReporter
import kafka.producer.{KeyedMessage, ProducerConfig, Producer}
import java.util

class TopicReporter(metricsRegistry: MetricsRegistry,
                    producerConfig: ProducerConfig,
                    prefix: String,
                    predicate: MetricPredicate = MetricPredicate.ALL,
                    clock: Clock = Clock.defaultClock()) extends AbstractPollingReporter(metricsRegistry, "kafka-topic-reporter") with MetricProcessor[Context] {
  private val producerMap: util.Map[MetricName, Producer[AnyRef, AnyRef]] = new util.HashMap[MetricName, Producer[AnyRef, AnyRef]]
  private var startTime: Long = 0L

  def run() {
    val metrics: util.Set[util.Map.Entry[MetricName, Metric]] = getMetricsRegistry.allMetrics.entrySet
    try {
      import scala.collection.JavaConversions._
      for (entry <- metrics) {
        val metricName: MetricName = entry.getKey
        val metric: Metric = entry.getValue
        if (predicate.matches(metricName, metric)) {
          val context: Context = new Context {
            def getProducer(header: String): Producer[AnyRef, AnyRef] = {
              val producer: Producer[AnyRef, AnyRef] = getActualProducer(metricName, header)
              producer
            }
          }
          metric.processWith(this, entry.getKey, context)
        }
      }
    }
    catch {
      case e: Exception => {
        e.printStackTrace
      }
    }
  }

  def processMeter(name: MetricName, meter: Metered, context: Context) {
    val header: String = "# time,count,1 min rate,mean rate,5 min rate,15 min rate"
    val producer: Producer[AnyRef, AnyRef] = context.getProducer(header)
    val topic = "%s-metrics-meter".format(prefix)
    val message = new StringBuilder().append(meter.count).append(',').append(meter.oneMinuteRate).append(',').append(meter.meanRate).append(',').append(meter.fiveMinuteRate).append(',').append(meter.fifteenMinuteRate).toString
    send(producer, header, topic, message)
  }

  def processCounter(name: MetricName, counter: Counter, context: Context) {
    val header: String = "# time,count"
    val producer: Producer[AnyRef, AnyRef] = context.getProducer(header)
    val topic = "%s-metrics-counter".format(prefix)
    val message = counter.count.toString
    send(producer, header, topic, message)
  }

  def processHistogram(name: MetricName, histogram: Histogram, context: Context) {
    val header: String = "# time,min,max,mean,median,stddev,95%,99%,99.9%"
    val producer: Producer[AnyRef, AnyRef] = context.getProducer(header)
    val snapshot: Snapshot = histogram.getSnapshot
    val topic = "%s-metrics-histogram".format(prefix)
    val message = new StringBuilder().append(histogram.min).append(',').append(histogram.max).append(',').append(histogram.mean).append(',').append(snapshot.getMedian).append(',').append(histogram.stdDev).append(',').append(snapshot.get95thPercentile).append(',').append(snapshot.get99thPercentile).append(',').append(snapshot.get999thPercentile).toString
    send(producer, header, topic, message)
  }

  def processTimer(name: MetricName, timer: Timer, context: Context) {
    val header: String = "# time,min,max,mean,median,stddev,95%,99%,99.9%"
    val producer: Producer[AnyRef, AnyRef] = context.getProducer(header)
    val snapshot: Snapshot = timer.getSnapshot
    val topic = "%s-metrics-timer".format(prefix)
    val message = new StringBuilder().append(timer.min).append(',').append(timer.max).append(',').append(timer.mean).append(',').append(snapshot.getMedian).append(',').append(timer.stdDev).append(',').append(snapshot.get95thPercentile).append(',').append(snapshot.get99thPercentile).append(',').append(snapshot.get999thPercentile).toString
    send(producer, header, topic, message)
  }

  def processGauge(name: MetricName, gauge: Gauge[_], context: Context) {
    val header: String = "# time,value"
    val producer: Producer[AnyRef, AnyRef] = context.getProducer(header)
    val topic = "%s-metrics-gauge".format(prefix)
    val message = gauge.value.toString
    send(producer, header, topic, message)
  }

  override def start(period: Long, unit: TimeUnit) {
    this.startTime = clock.time
    super.start(period, unit)
  }

  override def shutdown() {
    try {
      super.shutdown()
    }
    finally {
      import scala.collection.JavaConversions._
      for (producer <- producerMap.values) {
        producer.close()
      }
    }
  }

  private def getActualProducer(metricName: MetricName, header: String): Producer[AnyRef, AnyRef] = {
    var producer: Producer[AnyRef, AnyRef] = null
    producerMap synchronized {
      producer = producerMap.get(metricName)
      if (producer == null) {
        producer = new Producer(producerConfig)
        producerMap.put(metricName, producer)
      }
    }
    producer
  }

  private def send(producer: Producer[AnyRef, AnyRef], header: String, topic: String, message: String) = {
    val time: Long = TimeUnit.MILLISECONDS.toSeconds(clock.time - startTime)
    producer.send(new KeyedMessage(topic, "%s\n%d,%s".format(header, time, message).getBytes("UTF8")))
  }
}

trait Context {
  def getProducer(header: String): Producer[AnyRef, AnyRef]
}